storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -24,7 +26,178 @@ import (
+(START)-->|Job?|------>|CreatePV|-->|SetAvailable|--------->|Requeue|-->( STOP )
+ -----    +----+       +--------+   +------------+          +-------+    ------
+            | YES                                                ^
+            v                                                    |
+       +-----------+ Job Failed    +---------+                   |
+       |           |-------------->|SetFailed|------------------>+
+       |           |               +---------+                   |
+       |           |                                             |
+       |           | Unknown Job   +--------+                    |
+       |PollSaltJob|-------------->|UnsetJob|------------------->+
+       |           |               +--------+                    |
+       |           |                                             |
+       |           | Job Succeed   +--------+                    |
+       |           |-------------->|Job=DONE|------------------->+
+       +-----------+               +--------+                    |
+            | Job in progress                                    |
+            |                                                    |
+            +----------------------------------------------------+
+
+
+
+===================
+= Finalize Volume =
+===================
+
+`Pending` volumes cannot be deleted (because we don't know where we are in the
+creation process), so we reschedule the request until the volume becomes either
+`Failed` or `Available`.
+
+For volumes with no backing PersistentVolume we directly go reclaim the storage
+on the node and upon completion we remove our finalizer to let Kubernetes delete
+us.
+
+If we do have a backing PersistentVolume, we delete it (if it's not already in a
+terminating state) and watch for the moment when it becomes unused (this is done
+by rescheduling). Once the backing PersistentVolume becomes unused, we go
+reclaim its storage and remove the finalizers to let the object be deleted.
+
+  -----                                                            ------
+ (START)                                                          ( STOP )
+  -----                                                            ------
+    |                                                                ^
+    |                                                                |
+    v                                                                |
++--------+ YES                                                    +-------+
+|Pending?|------------------------------------------------------->|Requeue|
++--------+                                                        +-------+
+    | NO                                                             ^
+    v                                                                |
++--------+ NO         +----------------+ NO  +--------+              |
+| HasPv? |----------->|IsPvTerminating?|---->|DeletePV|------------->|
```

Comment:
If `HasPV` is `No`, then how can `IsPvTerminating` be true and how could we then call `DeletePV`?

Reply:
Good catch.

Forgot to update the branch's label when I reworked the flow chart.